main/test/ca/uwaterloo/flix/api/lsp/TestCompletionProvider.scala

```diff
@@ -0,0 +1,120 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.api.lsp
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.api.lsp.provider.CompletionProvider
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.Token
+import ca.uwaterloo.flix.language.ast.TypedAst.Root
+import ca.uwaterloo.flix.language.ast.shared.{Input, SecurityContext, Source}
+import ca.uwaterloo.flix.util.Options
+import org.scalatest.funsuite.AnyFunSuite
+
+class TestCompletionProvider extends AnyFunSuite  {
+  /**
+    * A list of programs to test invariants on.
+    */
+  private val programs = List(
+    s"""
+       |def main(): Unit \\ IO =
+       |    run {
+       |        Console.println("Please enter your name: ");
+       |        let name = Console.readln();
+       |        Console.println("Hello $${name}")
+       |    } with Console.runWithIO
+       |""".stripMargin,
+    s"""
+       |def main(): Unit \\ IO =
+       |    run {
+       |        let timestamp = Clock.currentTime(TimeUnit.Milliseconds);
+       |        println("$${timestamp} ms since the epoc")
+       |    } with Clock.runWithIO
+       |""".stripMargin
+  )
+
+  /**
+    * The uri of the test source.
+    */
+  private val uri = "<test>"
+
+  /**
+    * Compiles the given input string `s` with the given compilation options `o`.
+    */
+  private def compile(s: String, o: Options): (Root, Flix, List[CompilationMessage]) = {
+    implicit val sctx: SecurityContext = SecurityContext.AllPermissions
+    val flix = new Flix().setOptions(o).addSourceCode(uri, s)
+    flix.check() match {
+      case (Some(root), errors) => (root, flix, errors)
+      case (None, _) => fail("Compilation failed: a root is expected.")
+    }
+  }
+
+  /**
+    * Creates a source object from the given string `content`.
+    */
+  private def mkSource(content: String): Source = {
+    val sctx = SecurityContext.AllPermissions
+    val input = Input.Text(uri, content, sctx)
+    Source(input, content.toCharArray)
+  }
+
+  /**
+    * Checks the completion provider invariant for the given token filter, string transformer and completion assertion.
+    * For all the programs to be tested, we use the filter to select tokens to transform, apply the transformer to the selected tokens,
+    * call the completion provider with the transformed program and check the completions against the assertion.
+    *
+    * @param tokenFilter          the filter to select tokens to transform.
+    * @param stringTransformer    the transformer to apply to the selected tokens.
+    * @param completionAssertion  the assertion to check the completions against.
+    */
+  private def checkInvariant(tokenFilter: Token => Boolean, stringTransformer: String => String, completionAssertion: CompletionList => Unit ): Unit = {
```

Comment:
This is like a "god function". In OO one would say that it has too much responsibility.

Reply:
We need to break it down into smaller reusable parts.

Reply:
Will you break this down now or later?

Reply:
Don't add more higher order arguments. The opposite. You want to have small functions that each test case compose.